### ISSUE TITLE
invalid-type-form `ClassVar` cannot contain type variables

### ISSUE THREAD

**Issue by MarcelWilson** [astral-sh/ty#2828]
### Summary

I have a base dataclass that inserts each instance of the subclass into it's class var `all_self`.  
```python
@dataclass
class Base:
    all_self: ClassVar[list[Self]] = field(compare=False, repr=False)  # < -- ty error

    def __init_subclass__(cls: type[Self]) -> None:
        cls.all_self = []

    def setup_by(self: Self, actor: Actor) -> Self:
        reveal_type(self.all_self)
```
```
error[invalid-type-form] `ClassVar` cannot contain type variables
```

Is this a false positive with the `Self` annotation? 

mypy reveal_type:
```
Revealed type is "builtins.list[Self`0]" (30:20)
```

vs ty reveal_type:
```
info[revealed-type] Revealed type: `list[Self@setup_by]`
```

### Version

0.0.17

**Comment by carljm** [comment]
Yeah, other type checkers seem OK with this, so we probably need to make an exception for `Self` here. Maybe this should be clarified in the spec, too.

**Comment by AlexWaygood** [comment]
Do we know what the underlying principle is here for why other type checkers are okay with this, despite them disallowing type variables inside `Self` in general?

**Comment by carljm** [comment]
Can't speak for other type-checkers, but non-method `Self` is a bit weird, it's not a normal typevar. With a regular typevar `class Base[T]`, you can specialize to `Base[str]`, `Base[int]`, etc. So using the typevar in a classvar is problematic because the class itself can't be specialized with any particular specialization of `T`. (You could consider `T` specialized in a generic alias, which would suggest that regular typevars could be allowed in `ClassVar` -- but I don't think there's much appetite for that, since it wades into the quagmire that is "how you can use a generic alias at runtime", and would sort of require that each generic alias would have its own copy of the classvar, which is not how things work at runtime.)

This issue doesn't apply to `Self`: the only way you "specialize" `Self` in `Base` is by subclassing it. A class attribute `instances: ClassVar[list[Self]]` has an unambiguous interpretation: `Base.instances` is `list[Base]`, `Sub.instances` (given `class Sub(Base): ...`) is `list[Sub]`. And each subclass can have its own copy of the classvar at runtime.

Of course this interpretation is unsound because it ignores variance (if you pass the class `Sub` to somewhere expecting `t: type[Base]`, it'll let you unsoundly add a `Base` to `t.instances`). But that isn't specific to `ClassVar`, it's generally true of all non-method `Self`, so we've already accepted that unsoundness, for compatibility. (Though I'd like to add an opt-in diagnostic for it in future.)

**Comment by AlexWaygood** [comment]
Alright, that's good enough for me
